docs:add tut for postgres (#804)
bytebase/bytebase.com: content/blog/database-change-management-with-postgresql.md

---
title: Database Change Management with PostgreSQL
author: Ningjing
published_at: 2023/02/14 12:15
feature_image: /static/blog/database-change-management-with-postgresql/feature-image.webp
tags: How-To
description: PostgreSQL is the world's most advanced open-source relational database management system. This tutorial will guide you step-by-step to set up database change management for PostgreSQL in Bytebase.
---

This is a series of articles about Database Change Management with PostgreSQL.

- Database Change Management with PostgreSQL (this one)
- Database Change Management with PostgreSQL and GitHub

---

[PostgreSQL](https://www.postgresql.org/) is the world's most advanced open-source relational database management system.

This tutorial is a step-by-step guide to setting up **Database Change Management** for PostgreSQL in Bytebase. With Bytebase, a team can have a formalized review and rollout process to make PostgreSQL database schema change and data change.

> Bytebase provides a GUI for teams to perform **Database Changes** and retain full **Migration History**. Bytebase **Free Plan** is sufficient for this tutorial.

Reviewer: ```suggestion
Bytebase provides a GUI for teams to perform **Database Changes** and retain full **Change History**. Bytebase **Free Plan** is sufficient for this tutorial.
```

- Change Workflow
- Schema Editor
- SQL Advisor
- Change History
- SQL Editor Admin Mode
- Drift Detection

## Prerequisites
Before you start, make sure you have [Docker](https://www.docker.com/) installed.

## Step 1 - Deploy Bytebase via Docker

1. Make sure your Docker daemon is running, and start the Bytebase Docker container with following command:
  
````bash
docker run --init \
  --name bytebase \
  --platform linux/amd64 \
  --restart always \
  --publish 5678:8080 \
  --health-cmd "curl --fail http://localhost:5678/healthz || exit 1" \
  --health-interval 5m \
  --health-timeout 60s \
  --volume ~/.bytebase/data:/var/opt/bytebase \
  bytebase/bytebase:%%bb_version%% \
  --data /var/opt/bytebase \
  --port 8080
````

2. Bytebase is now running via Docker, and you can access it via `localhost:5678`.
![docker](/static/blog/database-change-management-with-postgresql/docker.webp)

1. Visit `localhost:5678` in your browser. Register the first admin account which will be granted [`Workspace Owner`](/docs/concepts/roles-and-permissions).
![register](/static/blog/database-change-management-with-postgresql/register.webp)

## Step 2 - Add PostgreSQL in Bytebase

> In Bytebase, ​​an Instance could be your on-premises MySQL instance, an AWS RDS instance etc, in this tutorial, a PostgreSQL instance.

Reviewer: ```suggestion
In Bytebase, ​​an Instance could be your on-premises PostgreSQL instance, an AWS RDS instance etc, in this tutorial.
```

1. Visit `localhost:5678` and log in.
![login](/static/blog/database-change-management-with-postgresql/login.webp)

> 2. Click **Instances** on the top bar you’ll see there is a sample PostgreSQL instance that is embedded. If you want to use your own instance, click **Add instance**. To keep it simple, we’ll use the sample one for this tutorial.

Reviewer: ```suggestion
2. Click **Instances** on the top bar you’ll see there is a sample PostgreSQL instance that is embedded. If you want to use your own instance, click **Add instance**. To keep it simple, we’ll use the built-in sample instance for this tutorial.
```

![bb-instances](/static/blog/database-change-management-with-postgresql/bb-instances.webp)

> ## Step 3 - Create a Project with PostgreSQL instance

Reviewer: ```suggestion
## Step 3 - Create a Project
```

3. **Title** and **SQL** will be filled automatically. You can select the **Assignee** to review the issue and then click **Create**.
![bb-issue-new-t1](/static/blog/database-change-management-with-postgresql/bb-issue-new-t1.webp)

4. Bytebase will run some basic checks and wait for the **Assignee** to approve.
![bb-issue-checks-approve](/static/blog/database-change-management-with-postgresql/bb-issue-checks-approve.webp)

5. Click **Approve** and the SQL will execute. Click **Resolve issue** and the issue status will become `Done`.
![bb-issue-t1-resolve](/static/blog/database-change-management-with-postgresql/bb-issue-t1-resolve.webp)

6. On the issue page, click **View change**. You will see the database difference.
![bb-view-change](/static/blog/database-change-management-with-postgresql/bb-view-change.webp)

7. You can also go to **Change History** under the project to view the full history. Or go into a specific database to view its history.
![bb-project-change-history](/static/blog/database-change-management-with-postgresql/bb-project-change-history.webp)

![bb-db-demo-change-history](/static/blog/database-change-management-with-postgresql/bb-db-demo-change-history.webp)

> ## Bonus Section - Drift Detect

Reviewer: ```suggestion
## Bonus Section - Schema Drift Detection
```

> This section requires you to have **Team Plan** or **Enterprise Plan** (you can start 14 days trial directly in the product without credit card).

Reviewer: ```suggestion
This section requires you to have **Enterprise Plan** (you can start 14 days trial directly in the product without credit card).
```

![bb-start-14-days-trial](/static/blog/database-change-management-with-postgresql/bb-start-14-days-trial.webp)

> Now you can see the full migration history of `demo`. However, what is Establish new baseline? When should it be used?

Reviewer: ```suggestion
Now you can see the full change history of `demo`. However, what is Establish new baseline? When should it be used?
```

2. Paste the following and then press **Enter**:

```sql
ALTER TABLE t1
ADD COLUMN age integer;`
```

3. Paste the following and then press **Enter** to verify it’s there:

```sql
SELECT 
  table_name,
  column_name,
  data_type
FROM
  information_schema.columns
WHERE
  table_name = 't1';
```
![bb-sql-admin-verify](/static/blog/database-change-management-with-postgresql/bb-sql-admin-verify.webp)
Admin mode skips Bytebase system and works as an external input. The change will take effect in Bytebase GUI in the next sync.
  
4. Wait for 10 mins. Go to **Anomaly Center**, and you can find the **Schema Drift**.
![bb-anomaly-center-1-drift](/static/blog/database-change-management-with-postgresql/bb-anomaly-center-1-drift.webp)

5. Click View diff, and you’ll see the drift.
![bb-drift-diff-age](/static/blog/database-change-management-with-postgresql/bb-drift-diff-age.webp)

6. You may also find the drift by clicking **Databases** > **demo**.
![bb-databases-demo-1-drift](/static/blog/database-change-management-with-postgresql/bb-databases-demo-1-drift.webp)

> 7. Go to **Databases** > **demo** > **Migration History** and click **Establish new baseline**.

Reviewer: ```suggestion
7. Go to **Databases** > **demo** > **Change History** and click **Establish new baseline**.
```

![bb-databases-demo-new-baseline](/static/blog/database-change-management-with-postgresql/bb-databases-demo-new-baseline.webp)

8. It will create an issue. Click **Approve** and **Resolve** to make it `Done`.
![bb-issue-baseline-done](/static/blog/database-change-management-with-postgresql/bb-issue-baseline-done.webp)

9. Go back to **Databases** > **demo** or **Anomaly Center**, and you will find the drift is gone.
![bb-databases-demo-0-drift](/static/blog/database-change-management-with-postgresql/bb-databases-demo-0-drift.webp)
![bb-anomaly-center-0-drift](/static/blog/database-change-management-with-postgresql/bb-anomaly-center-0-drift.webp)

## Summary and Next

> Now you have connected PostgreSQL with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full migration history for you. With **Team** or **Enterprise Plan**, you can even have drift detection.

Reviewer: ```suggestion
Now you have connected PostgreSQL with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full change history for you. With **Enterprise Plan**, you can even have schema drift detection.
```
